File: backend_python/src/todo_api/infrastructure/service/transaction_service.py
from collections.abc import Callable
from typing import TypeVar

from sqlalchemy.orm import Session

from todo_api.application_service.service.transaction_service import TransactionService
from todo_api.domain.repository.context_provider import ContextProvider

T = TypeVar("T")


class TransactionServiceImpl(TransactionService):
    def __init__(self, context_provider: ContextProvider[Session]) -> None:
        self.context_provider = context_provider

    def Run(self, func: Callable[[], T]) -> T:
        with self.context_provider.transaction():
            return func()

File: backend_bun/src/infrastructure/service/transactionService.ts
import type { SQL } from "bun";
import type { TransactionService } from "@/application_service/service/transactionService";
import type { ContextProvider } from "@/domain/repository/context";
import type { Conn } from "../repository/context";

export class TransactionServiceImpl implements TransactionService {
  constructor(
    private readonly db: SQL,
    private contextProvider: ContextProvider<Conn>,
  ) {}
  async run<T>(fn: () => Promise<T>): Promise<T> {
    return this.db.begin((tx) => {
      return this.contextProvider.runWith(tx, fn);
    });
  }
}
